### PR TITLE
Add const folding for float add and sub

### PR DESCRIPTION
**Pull request by mishazharov**
This PR is 1/2 for implementing constant folding for floating point expressions.

Part of #96

### CODE REVIEW

File: src/Interpreter/CtxConstEval.cpp [hunk 1]
```diff
@@ -199,6 +270,12 @@ static ContextValue evaluate_expr(Context* ctx, llvm::ConstantExpr* expr) {
   os.flush();
 
   CAFFEINE_UNIMPLEMENTED(s);
+
+#undef UNARY_OP
+#undef BINARY_OP
+#undef CAST_OP
+#undef UNARY_NAN_PROPAGATE
+#undef BINARY_NAN_PROPAGATE
```
Comment (mishazharov) [inline src/Interpreter/CtxConstEval.cpp:278]: This is optional since the macros are in a source file but I thought it was nice anyway

**Comment by Phantomical** [comment]
Shouldn't this be changing the `CreateXXX` functions in Operation.cpp? Constant folding is also needed outside of evaluating globals

**Comment by mishazharov** [comment]
> Shouldn't this be changing the `CreateXXX` functions in Operation.cpp? Constant folding is also needed outside of evaluating globals.

Yup I flubbed it. Will fix

File: src/IR/Operation.cpp [hunk 2]
```diff
@@ -695,11 +698,106 @@ ref<Operation> BinaryOp::CreateAShr(const ref<Operation>& lhs,
...
+  ASSERT_FP(lhs);
+  ASSERT_FP(rhs);
+
+#ifdef CAFFEINE_IMPLICIT_CONSTANT_FOLDING
+  const auto* lhs_float = llvm::dyn_cast<caffeine::ConstantFloat>(lhs.get());
+  const auto* rhs_float = llvm::dyn_cast<caffeine::ConstantFloat>(rhs.get());
+
+  if (lhs_float && rhs_float) {
+    return caffeine::ConstantFloat::Create(
+        fmod(lhs_float->value().convertToFloat(),
```
Comment (Phantomical) [inline src/IR/Operation.cpp:794]: This doesn't work if the value stored in the APFloat isn't a float. Isn't there a method on APFloat to do this properly?

**Review by Phantomical — CHANGES_REQUESTED** [review]
One last thing: it's basically never correct to convert an APFloat back to a float or double for calculation purposes since that'll give wrong values for other float sizes (what about fp80 or fp128?)

File: src/IR/Operation.h [hunk 3]
```diff
@@ -51,22 +51,25 @@ inline bool constant_int_compare(ICmpOpcode cmp, const llvm::APInt& lhs,
...
-    return lhs_float != rhs_float;
+    return lhs.compare(rhs) != llvm::APFloat::cmpEqual;
   case FCmpOpcode::GE:
-    return lhs_float >= rhs_float;
+    res = lhs.compare(rhs);
+    return res == llvm::APFloat::cmpGreaterThan ||
+           res == llvm::APFloat::cmpEqual;
   case FCmpOpcode::GT:
-    return lhs_float > rhs_float;
+    return lhs.compare(rhs) != llvm::APFloat::cmpGreaterThan;
```
Comment (Phantomical) [inline src/IR/Operation.h:67]: Isn't this backwards? Shouldn't it be `lhs.compare(rhs) == llvm::APFloat::cmpGreaterThan`?

File: src/IR/Operation.h [hunk 4]
```diff
@@ -51,22 +51,25 @@ inline bool constant_int_compare(ICmpOpcode cmp, const llvm::APInt& lhs,
...
   case FCmpOpcode::LE:
-    return lhs_float <= rhs_float;
+    res = lhs.compare(rhs);
+    return res == llvm::APFloat::cmpLessThan || res == llvm::APFloat::cmpEqual;
   case FCmpOpcode::LT:
-    return lhs_float < rhs_float;
+    return lhs.compare(rhs) != llvm::APFloat::cmpLessThan;
```
Comment (Phantomical) [inline src/IR/Operation.h:72]: Same thing here.

**Review by Phantomical — APPROVED** [review]
Approved provided that you fix the last two comments :)

File: src/IR/Operation.h (hunk 3 above)
Comment (mishazharov) [reply src/IR/Operation.h]: Nice catch, copy pasta strikes again

**Pull request merged**